### DISCUSSION TITLE
Warning: Function components cannot be given refs

### DISCUSSION THREAD

**Discussion by nareshbhatia** [react-hook-form/react-hook-form#6256] [Q&A]
I am trying to integrate the `TextInput` component from the [USDSW library](https://trussworks.github.io/react-uswds/?path=/docs/components-form-controls-text-input--default-text-input) with RHF. This component has a `ref` property. Everything is working fine, but I am getting the following error in Chrome DevTools console:

> Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?

How can I get rid of this warning? Please see [my CodeSandbox here](https://codesandbox.io/s/react-hook-form-with-ui-library-forked-i87ue).

```tsx
const TextFieldUswds = React.forwardRef(
  ({ id, name, label, error, onBlur, onChange }, ref) => {
    return (
      <Fragment>
        {label !== undefined ? <label htmlFor={id}>{label}</label> : null}
        <TextInput
          id={id}
          name={name}
          type="text"
          ref={ref}
          className="usa-input"
          onBlur={onBlur}
          onChange={onChange}
        />
        <ErrorMessage>{error}</ErrorMessage>
      </Fragment>
    );
  }
);

const schema = yup.object().shape({
  firstName: yup.string().required(),
  lastName: yup.string().required()
});

function TestForm({ onSubmit }) {
  const methods = useForm({
    mode: "onBlur",
    resolver: yupResolver(schema)
  });
  const { formState, register, handleSubmit } = methods;
  const { errors } = formState;

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className="margin-bottom-2">
          <TextFieldUswds
            id="firstName"
            {...register("firstName")}
            label="First Name"
            error={errors.firstName?.message}
          />
        </div>

        <div className="margin-bottom-2">
          <TextFieldUswds
            id="lastName"
            {...register("lastName")}
            label="Last Name"
            error={errors.lastName?.message}
          />
        </div>

        <Button type="submit">Submit</Button>
      </form>
    </FormProvider>
  );
}
```

**Comment by nareshbhatia** [comment]
I think I may have solved it, but don't understand exactly how. In the code below, I simply replaced `ref={ref}` with `inputRef={ref}`.

```tsx
<TextInput
  id={id}
  name={name}
  type="text"
  ref={ref}
  className="usa-input"
  onBlur={onBlur}
  onChange={onChange}
/>
```

The library provides both props, but I don't understand the difference:

<img width="689" alt="CleanShot 2021-08-18 at 10 27 19@2x" src="https://user-images.githubusercontent.com/1220198/129916566-5d973420-0ef3-40a0-929c-f90a09153740.png">

**Reply by bluebill1049** [reply]
react hook form wants the input reference instead of the entire dom reference, but the next minor version will resolve this problem.

**Reply by nareshbhatia** [reply]
So what's the difference between "input reference vs the entire dom reference". Not sure if I understand.

**Reply by bluebill1049** [reply]
`<input />`

vs

```
<div>
  <input />
</div>
```
